### PR TITLE
SAML: Support for authnContextClassRef in proxy mode

### PR DESCRIPTION
**Pull request by tauceti2**
If SSP works as a proxy, this code retain requested authnContextClassRef from
SP behind the proxy and pass it to the upper IdP. It also pass the
authnContextClassRef sent by the upper IdP back to the SP behind the proxy. By this change SPs behind the proxy can request different authn methods, especially useful for support multi-factor authn.

Please review the code, if it is acceptable also for mode when SSP doesn't work as a proxy.

### CODE REVIEW

**Comment by pradtke** [comment]
We have some proxy use cases where this is undesirable so it would be nice if there was a way to opt in or opt out of the behavior change.  

Use case is that SPs request MFA from the proxy. The proxy performs it own MFA for users and we don't want to request MFA from upstream IdPs.

**Comment by tauceti2** [comment]
Hi @pradtke, added configuration option, which can enable that behaviour. By default it is off.

**Review by dmotelica — CHANGES_REQUESTED** [review]
Hello,
For those who use SimpleSAMLphp as a proxy to delegate l'auhentication to Azure AD and have to be NIH ready (Refeds MFA profile):

File SP.php - add this before line 487 :
--------------------
`if ($state['saml:RequestedAuthnContext']['AuthnContextClassRef'][0]=="https://refeds.org/profile/mfa"){
					$state['saml:RequestedAuthnContext']['AuthnContextClassRef'][0] = 'http://schemas.microsoft.com/claims/multipleauthn';
				}	`

----------------------------------------------------------------

file SAML2.php -  - add this before line 1137:
-----------------------------------------------
`if(isset($state['saml:RequestedAuthnContext']) && $state['saml:sp:AuthnContext']=="http://schemas.microsoft.com/claims/multipleauthn") 
			{
				$state['saml:sp:AuthnContext']="https://refeds.org/profile/mfa";
			}` 

---------------------------------------------------------

in config.php and authsources.php 
------------------------------------
`'proxymode.passAuthnContextClassRef' => true,`

**Comment by tvdijen** [comment]
Way too specific to add that @dmotelica ..

File: modules/saml/lib/Auth/Source/SP.php [hunk 1]
```diff
@@ -460,6 +476,19 @@ private function startSSO2(Configuration $idpMetadata, array $state): void
...
             $ar->setRequestedAuthnContext(['AuthnContextClassRef' => $accr, 'Comparison' => $comp]);
+        } elseif ($this->passAuthnContextClassRef && isset($state['saml:RequestedAuthnContext']) && isset($state['saml:RequestedAuthnContext']['AuthnContextClassRef'])) {
+            if (isset($state['saml:RequestedAuthnContext']['Comparison']) && in_array($state['saml:RequestedAuthnContext']['Comparison'], array(
+                                                SAML2\Constants::COMPARISON_EXACT,
+                                                SAML2\Constants::COMPARISON_MINIMUM,
+                                                SAML2\Constants::COMPARISON_MAXIMUM,
+                                                SAML2\Constants::COMPARISON_BETTER,
+                        ))) {
+                // RequestedAuthnContext has been set by an SP behind the proxy so pass it to the upper IdP
+                $ar->setRequestedAuthnContext(array(
```
Comment (dmotelica) [inline modules/saml/lib/Auth/Source/SP.php:487]: `if($state['saml:RequestedAuthnContext']['AuthnContextClassRef'][0]=="https://refeds.org/profile/mfa"){
 $state['saml:RequestedAuthnContext']['AuthnContextClassRef'][0] = 
'http://schemas.microsoft.com/claims/multipleauthn'; }`

**Comment by ghalse** [comment]
This pull request is now referenced from the [REFEDS MFA Profile implementation FAQ](https://wiki.refeds.org/display/PRO/IdP+Product+specific+questions#IdPProductspecificquestions-IrunSimpleSAMLPHPIdP,whattoIneedtodotosupportREFEDSMFAProfile).

For that reason, I wanted to note that since the patch was first written there are changes to SSP that means you need to make sure you include the later merge commits (and particularly [2c6cf5a](https://github.com/simplesamlphp/simplesamlphp/pull/833/commits/2c6cf5a7ed39b6f9270392e9434ccbe1472a37c1)) if you're trying to use this as a patch against recent SSP.

**Comment by ghalse** [comment]
I've forward ported this for simplesamlphp 2.0.0-rc2, and the patch is in 554466328f388a6bcc686d5ce6df6f607d1a352f

**Comment by tvdijen** [comment]
I can't seem to cherry-pick that commit.. It's not part of any branch.. What exactly did you do @ghalse ?

**Comment by ghalse** [comment]
> I can't seem to cherry-pick that commit.. It's not part of any branch.. What exactly did you do @ghalse ?

It's part of the authnContextClassRef branch in ghalse/simplesamlphp, which was based off the simplesamlphp-2.0 branch

https://github.com/simplesamlphp/simplesamlphp/compare/simplesamlphp-2.0...ghalse:simplesamlphp:authnContextClassRef

**Comment by tvdijen** [comment]
Ah, thanks! Cherry-pick it!

**Pull request closed without merging**